src/services/pure-di.ts

```diff
@@ -0,0 +1,63 @@
+import axios, { AxiosStatic } from 'axios';
+import { HttpDopplerLegacyClient, DopplerLegacyClient } from './doppler-legacy-client';
+import { OnlineSessionManager, SessionManager } from './session-manager';
+
+interface AppConfiguration {
+  dopplerLegacyUrl: string;
+  dopplerLegacyKeepAliveMilliseconds: number;
+}
+
+export interface AppContainer {
+  axiosStatic: AxiosStatic;
+  appConfiguration: AppConfiguration;
+  dopplerLegacyClient: DopplerLegacyClient;
+  sessionManager: SessionManager;
+}
+
+export class AppCompositionRoot implements AppContainer {
+  private readonly instances: Partial<AppContainer>;
+
+  constructor(predefinedInstances: Partial<AppContainer> = {}) {
+    this.instances = { ...predefinedInstances };
+  }
+
+  private singleton<N extends keyof AppContainer, T extends AppContainer[N]>(
```

Comment:
This type is not so easy to read, but it is great!:

It only allows call `Singleton` function with the right key and type pair, for example: 

```
this.singleton('appConfiguration', X);
```

_X_ could be only a object that implements the class _AppConfiguration_